pkg/types/validation/installconfig.go

```diff
@@ -99,6 +102,122 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
 	if _, ok := validPublishingStrategies[c.Publish]; !ok {
 		allErrs = append(allErrs, field.NotSupported(field.NewPath("publish"), c.Publish, validPublishingStrategyValues))
 	}
+
+	return allErrs
+}
+
+// IPAddressTypeByField is a map of field path to whether they request IPv4 or IPv6.
+type IPAddressTypeByField map[string]struct{ IPv4, IPv6 bool }
+
+// IPByField is a map of field path to the net.IPs in sorted order.
+type IPByField map[string][]net.IP
+
+// InferIPVersionFromInstallConfig infers the user's desired ip version from the networking config.
+// Presence field names match the field path of the struct within the Networking type. This function
+// assumes a valid install config.
+func InferIPVersionFromInstallConfig(n *types.Networking) (hasIPv4, hasIPv6 bool, presence IPAddressTypeByField, addresses IPByField) {
+	if n == nil {
+		return
+	}
+	addresses = make(IPByField)
+	for _, network := range n.MachineNetwork {
+		addresses["machineNetwork"] = append(addresses["machineNetwork"], network.CIDR.IP)
+	}
+	for _, network := range n.ServiceNetwork {
+		addresses["serviceNetwork"] = append(addresses["serviceNetwork"], network.IP)
+	}
+	for _, network := range n.ClusterNetwork {
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+	}
+	presence = make(IPAddressTypeByField)
+	for k, ips := range addresses {
+		sort.Slice(ips, func(i, j int) bool { return bytes.Compare(ips[i], ips[j]) < 0 })
```

Comment:
can we move this burden to the tests if its requirement for that? because i don't know why we are sorting these?

Reply:
For printing to users.  So they’re stable.

Reply:
if they are only for printing maybe sort the strings in https://github.com/openshift/installer/pull/2829/files/e953538d7c6f35033fadd216bf06a91d9a8753dc#diff-02200a3eac2ac223b4117e690cf2d0c7R150-R156

i think string sort is more easier to reason about comapred to https://github.com/openshift/installer/pull/2829/files/e953538d7c6f35033fadd216bf06a91d9a8753dc#r358821380

Reply:
ok

Reply:
I actually removed this, I want them in the order the user specifies anyway.